source/h2_decoder.c

```diff
@@ -718,8 +737,11 @@ static int s_state_fn_frame_settings_loop(struct aws_h2_decoder *decoder, struct
     (void)input;
 
     if (decoder->frame_in_progress.payload_len == 0) {
-        /* Huzzah, done with the frame */
-        DECODER_CALL_VTABLE(decoder, on_settings_end);
+        /* Huzzah, done with the frame, fire the callback */
+        struct aws_array_list *buffer = &decoder->settings_buffer_list;
+        DECODER_CALL_VTABLE_ARGS(decoder, on_settings, buffer, aws_array_list_length(&decoder->settings_buffer_list));
+        /* clean up the buffer */
+        aws_array_list_clean_up(&decoder->settings_buffer_list);
```

Comment:
BUG: You're being careful to clean_up the settings_buffer_list each time we have an error related to settings, but what if we need to shut down for some reason unrelated to the decoder? we forgot to aws_array_list_clean_up() in aws_h2_decoder_destroy()

It would be less error-prone to just clean up the array **once**, in aws_h2_decoder_destroy. 

We can call aws_array_list_clear() here, after delivering the settings to the connection, to reset its length to 0.